components/client/typescript/src/server.ts

```diff
@@ -0,0 +1,187 @@
+import { Predicate, PredicateHeaderSchema, ThenThatHttpPost } from './schemas/predicate';
+import { BitcoinIfThisOptionsSchema, BitcoinIfThisSchema } from './schemas/bitcoin/if_this';
+import { StacksIfThisOptionsSchema, StacksIfThisSchema } from './schemas/stacks/if_this';
+
+/** Function type for a Chainhook event callback */
+export type OnEventCallback = (uuid: string, payload: Payload) => Promise<void>;
+
+const ServerOptionsSchema = Type.Object({
+  hostname: Type.String(),
+  port: Type.Integer(),
+  auth_token: Type.String(),
+  external_host: Type.String(),
+});
+/** Local event server connection and authentication options */
+export type ServerOptions = Static<typeof ServerOptionsSchema>;
+
+const ChainhookNodeOptionsSchema = Type.Object({
+  hostname: Type.String(),
+  port: Type.Integer(),
+});
+/** Chainhook node connection options */
+export type ChainhookNodeOptions = Static<typeof ChainhookNodeOptionsSchema>;
+
+const IfThisThenNothingSchema = Type.Union([
+  Type.Composite([
+    BitcoinIfThisOptionsSchema,
+    Type.Object({
+      if_this: BitcoinIfThisSchema,
+    }),
+  ]),
+  Type.Composite([
+    StacksIfThisOptionsSchema,
+    Type.Object({
+      if_this: StacksIfThisSchema,
+    }),
+  ]),
+]);
+const ServerPredicateSchema = Type.Composite([
+  PredicateHeaderSchema,
+  Type.Object({
+    networks: Type.Union([
+      Type.Object({
+        mainnet: IfThisThenNothingSchema,
+      }),
+      Type.Object({
+        testnet: IfThisThenNothingSchema,
+      }),
+    ]),
+  }),
+]);
+/** Chainhook predicates registerable by the local event server */
+export type ServerPredicate = Static<typeof ServerPredicateSchema>;
+
+/**
+ * Build the Chainhook Fastify event server.
+ * @param serverOpts - Server options
+ * @param chainhookOpts - Chainhook node options
+ * @param predicates - Predicates to register
+ * @param callback - Event callback function
+ * @returns Fastify instance
+ */
+export async function buildServer(
+  serverOpts: ServerOptions,
+  chainhookOpts: ChainhookNodeOptions,
+  predicates: [ServerPredicate],
+  callback: OnEventCallback
+) {
+  const base_path = `http://${chainhookOpts.hostname}:${chainhookOpts.port}`;
+
+  async function waitForNode(this: FastifyInstance) {
+    logger.info(`EventServer connecting to chainhook node at ${base_path}...`);
+    while (true) {
+      try {
+        await request(`${base_path}/ping`, { method: 'GET', throwOnError: true });
+        break;
+      } catch (error) {
+        logger.error(error, 'Chainhook node not available, retrying...');
+        await timeout(1000);
+      }
+    }
+  }
+
+  async function registerPredicates(this: FastifyInstance) {
+    logger.info(predicates, `EventServer registering predicates at ${base_path}...`);
+    for (const predicate of predicates) {
+      const thenThat: ThenThatHttpPost = {
+        http_post: {
+          url: `http://${serverOpts.external_host}/chainhook/${predicate.uuid}`,
```

Comment:
Can we add https to these or does that not make sense -- this is not on the same machine, right?

Reply:
This could be configurable yes

Reply:
Good catch... changed the option so it includes the full base URL